src/graphene.cpp

```diff
@@ -0,0 +1,1504 @@
+    ostringstream ss;
+    ss << "Inbound CMemPoolInfo size (last 24hrs) AVG: " << formatInfoUnit(avgMemPoolInfoSize);
+    return ss.str();
+}
+
+// Calculate the average inbound graphene CMemPoolInfo size
+string CGrapheneBlockData::OutBoundMemPoolInfoToString()
+{
+    LOCK(cs_graphenestats);
+    double avgMemPoolInfoSize = average(mapMemPoolInfoOutBound);
+    ostringstream ss;
+    ss << "Outbound CMemPoolInfo size (last 24hrs) AVG: " << formatInfoUnit(avgMemPoolInfoSize);
+    return ss.str();
+}
+
+// Calculate the graphene percentage compression over the last 24 hours
+string CGrapheneBlockData::ResponseTimeToString()
+{
+    LOCK(cs_graphenestats);
+
+    vector<double> vResponseTime;
+
+    double nResponseTimeAverage = 0;
+    double nPercentile = 0;
+    double nTotalResponseTime = 0;
+    double nTotalEntries = 0;
+    for (map<int64_t, double>::iterator mi = mapGrapheneBlockResponseTime.begin();
+         mi != mapGrapheneBlockResponseTime.end(); ++mi)
+    {
+        nTotalEntries += 1;
+        nTotalResponseTime += (*mi).second;
+        vResponseTime.push_back((*mi).second);
+    }
+
+    if (nTotalEntries > 0)
+    {
+        nResponseTimeAverage = (double)nTotalResponseTime / nTotalEntries;
+
+        // Calculate the 95th percentile
+        uint64_t nPercentileElement = static_cast<int>((nTotalEntries * 0.95) + 0.5) - 1;
+        sort(vResponseTime.begin(), vResponseTime.end());
+        nPercentile = vResponseTime[nPercentileElement];
+    }
+
+    ostringstream ss;
+    ss << fixed << setprecision(2);
+    ss << "Response time   (last 24hrs) AVG:" << nResponseTimeAverage << ", 95th pcntl:" << nPercentile;
+    return ss.str();
+}
+
+// Calculate the graphene percentage compression over the last 24 hours
+string CGrapheneBlockData::ValidationTimeToString()
+{
+    LOCK(cs_graphenestats);
+
+    vector<double> vValidationTime;
+
+    double nValidationTimeAverage = 0;
+    double nPercentile = 0;
+    double nTotalValidationTime = 0;
+    double nTotalEntries = 0;
+    for (map<int64_t, double>::iterator mi = mapGrapheneBlockValidationTime.begin();
+         mi != mapGrapheneBlockValidationTime.end(); ++mi)
+    {
+        nTotalEntries += 1;
+        nTotalValidationTime += (*mi).second;
+        vValidationTime.push_back((*mi).second);
+    }
+
+    if (nTotalEntries > 0)
+    {
+        nValidationTimeAverage = (double)nTotalValidationTime / nTotalEntries;
+
+        // Calculate the 95th percentile
+        uint64_t nPercentileElement = static_cast<int>((nTotalEntries * 0.95) + 0.5) - 1;
+        sort(vValidationTime.begin(), vValidationTime.end());
+        nPercentile = vValidationTime[nPercentileElement];
+    }
+
+    ostringstream ss;
+    ss << fixed << setprecision(2);
+    ss << "Validation time (last 24hrs) AVG:" << nValidationTimeAverage << ", 95th pcntl:" << nPercentile;
+    return ss.str();
+}
+
+// Calculate the graphene percentage compression over the last 24 hours
```

Comment:
stale comment. the code compute the transaction re-request ratio and keep track of the number of the total re-request.

Reply:
Done.